### PR TITLE
[PHI] Fix performance issue in bilinear interpolation's backward kernel.

### PR DESCRIPTION
**Pull request by A-nnonymous**
### PR Category
Performance Optimization  

### PR Types
Performance 

### Description
Due to the excessively dense atomic operations used in the original backpropagation operator for bilinear interpolation, the performance deteriorates drastically under high parallelism caused by atomic operation contention. Through extensive quantitative analysis, this work designs a high-performance backpropagation operator upon the base of original operator that circumvents atomic operations and activates it under a certain threshold.

In typical cases, the optimized operator performs more than 3x better than the original operator, and its speedup ratio increases with the interpolation factor.

pcard-76996

### CODE REVIEW

File: paddle/phi/kernels/gpu/interpolate_grad_kernel.cu [hunk 1]
```diff
@@ -35,12 +35,13 @@ __forceinline__ __device__ void PreCalculatorForLinearInterpInputIndex(
...
     T src_x,
     const int in_img_x) {
-  src_x = (src_x > static_cast<T>(0)) ? src_x : static_cast<T>(0);
-  *in_img_idx = static_cast<int>(src_x);
-  *x_id = (*in_img_idx < in_img_x - 1) ? 1 : 0;
-  using MT = typename phi::dtype::MPTypeTrait<T>::Type;
-  *lambda1 = static_cast<T>(static_cast<MT>(src_x) - *in_img_idx);
-  *lambda2 = static_cast<T>(1.0) - *lambda1;
+  src_x = max(src_x, static_cast<T>(0));
+  T src_x_floor = floorf(src_x);
```
Comment (Hongqing-work) [inline paddle/phi/kernels/gpu/interpolate_grad_kernel.cu:39]: 这里max和floorf是否兼容所有type可以再确认下

File: paddle/phi/kernels/gpu/interpolate_grad_kernel.cu [hunk 2]
```diff
@@ -360,42 +361,115 @@ __global__ void KeBilinearInterpNCHWBw(T* in,
...
 
-  for (; index < num_out; index += stride) {
-    int index_tmp = index;
-    int w2 = index_tmp % out_w;
-    index_tmp /= out_w;
-    int h2 = index_tmp % out_h;
-    int nc = index_tmp / out_h;
+  // Restricted parallelism if ratio_w is over threshold
+  // to avoid atomic contention overhead.
+  if (ratio_w < 0.5) [[likely]] {  // NOLINT
```
Comment (Hongqing-work) [inline paddle/phi/kernels/gpu/interpolate_grad_kernel.cu]: 0.5f?

Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu]: 这个是后面测试出来的一个阈值，小于0.5时会有较大的原子操作竞争，可见文档https://ku.baidu-int.com/knowledge/HFVrC7hq1Q/pKzJfZczuc/X_AiPuGHxP/vurMXtRr-m4JxV

File: paddle/phi/kernels/gpu/interpolate_grad_kernel.cu (hunk 1 above)
Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu:39]: 感谢，确实可能存在兼容问题，修改中

Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu:39]: 经检查，src_x以及两个lambda在调用时使用的是MT，我目前发现只有float和double两种可能，具体可见/work/Paddle/paddle/phi/common/amp_type_traits.h

Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu:39]: 这个函数在循环很内部，对性能提升幅度的影响可能会达到10-20%，但是MT这块不知道会不会变，所以这个优化咱可以继续讨论下需不需要加

File: paddle/phi/kernels/gpu/interpolate_grad_kernel.cu (hunk 2 above)
Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu]: 数据格式问题，注意到了，感谢

Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu]: done

**Review by phlrain — APPROVED** [review]

File: paddle/phi/kernels/gpu/interpolate_grad_kernel.cu (hunk 1 above)
Comment (A-nnonymous) [reply paddle/phi/kernels/gpu/interpolate_grad_kernel.cu:39]: 经讨论，MT在短时间内不会改变，该函数的所有调用情况中，T仅可能为float或double，该问题可以不用考虑

**Review by Hongqing-work — APPROVED** [review]

**Pull request merged**